os/kernel/pthread/pthread_keydelete.c

```diff
@@ -111,13 +79,36 @@ int pthread_key_delete(pthread_key_t key)
 {
 	struct pthread_tcb_s *rtcb = (FAR struct pthread_tcb_s *)this_task();
 	struct task_group_s *group = rtcb->cmn.group;
+	struct pthread_key_s *cur_key;
 
 	DEBUGASSERT(group && (rtcb->cmn.flags & TCB_FLAG_TTYPE_MASK) == TCB_FLAG_TTYPE_PTHREAD);
 
-	if (key < PTHREAD_KEYS_MAX && group->tg_keys[key] == IN_USE) {
-		rtcb->pthread_data[key].data = NULL;
-		rtcb->pthread_data[key].destructor = NULL;
-		group->tg_keys[key] = NOT_IN_USE;
+	/* Check key validation */
+
+	if (key < PTHREAD_KEYS_MAX) {
+
+		/* Find key data */
+
+		cur_key = (struct pthread_key_s *)pthread_key_find(rtcb, key);
+		if (cur_key == NULL) {
+
+			/* No Key in the list */
+
+			return EINVAL;
+		}
+
+		/* Remove key from list */
+
+		sq_rem((sq_entry_t *)cur_key, &rtcb->key_list);
+
+		/* Free key data */
```

Comment:
Releasing the cur_key , without calling the corresponding destructor wouldn't cause issue ?

Reply:
Don't we need to reset the cur_key values before calling sched_kfree ?
		cur_key->data = NULL;
		cur_key->destructor = NULL;

Reply:
1. Pthread executes destructor on exit, not delete.
```
An optional destructor function may be associated with each key value. At thread exit, if a key value has a non-NULL destructor pointer, and the thread has a non-NULL value associated with that key, the value of the key is set to NULL, and then the function pointed to is called with the previously associated value as its sole argument.
```
From POSIX, ```at thread exit``` is written.

2. There are two reasons.
    1. User can not acess key->data and key->destructor directly. Pthread uses them internally.
    2. Pthread can acess key structure through key single list. Because pthread remove deleted key from list, set to NULL is unnecessary.